content/scripts/build.js

```diff
@@ -1234,12 +1281,159 @@ class Builder {
   excludeSlug(url) {
     // XXX would it be faster to compute a regex in the constructor
     // and use it repeatedly here instead?
+    const urlLC = url.toLowerCase();
     if (this.options.slugsearch.length) {
-      return !this.options.slugsearch.some((search) => url.includes(search));
+      return !this.options.slugsearch.some((search) => urlLC.includes(search));
     }
     return false;
   }
 
+  /**
+   * Recursive method that renders the macros within the document
+   * represented by this source, URI, metadata and raw HTML, as
+   * well as builds any live samples within this document or within
+   * other documents referenced by this document.
+   */
+  async renderMacrosAndBuildLiveSamples(
+    source,
+    uri,
+    metadata,
+    rawHtml,
+    destinationDir,
+    { cacheResult = false, selectedSampleIDs = null } = {}
+  ) {
+    // First, render the macros to get the rendered HTML.
+    const [renderedHtml, flaws] = await this.renderMacros(
+      source,
+      rawHtml,
+      metadata,
+      {
+        cacheResult,
+      }
+    );
+
+    // Next, let's find any samples that we need to build, and note
+    // that one or more or even all might be within other documents.
+    let ownSampleIds;
+    let otherSampleIds = null;
+    const liveSamplesDir = destinationDir + "$samples";
+    if (selectedSampleIDs) {
+      ownSampleIds = selectedSampleIDs;
+    } else {
+      [ownSampleIds, otherSampleIds] = kumascript.getLiveSampleIDs(
+        metadata.slug,
+        rawHtml
+      );
+    }
+
+    // Next, let's build the live sample pages from the current document, if any.
+    if (ownSampleIds) {
+      // First, remove any live samples that no longer exist.
+      if (fs.existsSync(liveSamplesDir)) {
+        for (const de of fs.readdirSync(liveSamplesDir, {
+          withFileTypes: true,
+        })) {
+          // All directories under the main live-samples directory for this
+          // document are specific live samples. If one exists that is not
+          // contained in the current set of live samples, remove it.
+          if (de.isDirectory() && !ownSampleIds.has(de.name)) {
+            rimraf.sync(path.join(liveSamplesDir, de.name));
+          }
+        }
+      }
+      // Now, we'll either update or create new live sample pages.
+      let liveSamplePageHtml;
+      for (const sampleID of ownSampleIds) {
+        try {
+          liveSamplePageHtml = kumascript.buildLiveSamplePage(
+            uri,
+            metadata.title,
+            renderedHtml,
+            sampleID
+          );
+        } catch (e) {
+          if (e instanceof kumascript.KumascriptError) {
+            flaws.push(e);
```

Comment:
This is not great. What happens is this:
<img width="997" alt="Screen Shot 2020-05-19 at 1 15 50 PM" src="https://user-images.githubusercontent.com/26739/82357290-e5742f00-99d2-11ea-8cbc-0deaf763a7c0.png">

What I did was that I was in content/files/en-us/web/api/canvas_api/tutorial/applying_styles_and_colors/index.html and I messed it up from...
```
57| <p>{{EmbedLiveSample("A_fillStyle_example", 160, 160, "https://mdn.mozillademos.org/files/5417/Canvas_fillstyle.png")}}</p>
```
...to...
```
57| <p>{{EmbedLiveSample("A_fillStyle_exampleXXX", 160, 160, "https://mdn.mozillademos.org/files/5417/Canvas_fillstyle.png")}}</p>
```

What I should get is an "earlier" error. I.e. the error should have bubbled up here to "Document A" (using the lingo we've been using). It should inform me that *I* have a problem here on line 57.

Reply:
I agree, but this is not a quick fix. I need to think about the best way to do that.